src/policy/immix/immixspace.rs

```diff
@@ -511,24 +566,36 @@ impl<VM: VMBinding> ImmixSpace<VM> {
         } else {
             // We won the forwarding race; actually forward and copy the object if it is not pinned
             // and we have sufficient space in our copy allocator
-            let new_object = if self.is_pinned(object) || self.defrag.space_exhausted() {
+            let new_object = if self.is_pinned(object)
+                || (!nursery_collection && self.defrag.space_exhausted())
+            {
                 self.attempt_mark(object, self.mark_state);
                 ForwardingWord::clear_forwarding_bits::<VM>(object);
                 Block::containing::<VM>(object).set_state(BlockState::Marked);
                 object
             } else {
-                ForwardingWord::forward_object::<VM>(object, semantics, copy_context)
+                let new_object =
+                    ForwardingWord::forward_object::<VM>(object, semantics, copy_context);
+                Block::containing::<VM>(new_object).set_state(BlockState::Marked);
```

Comment:
Both the `if` and the `else` branches set the marked state.  We may merge them and move them after the `if` statement.

Reply:
I agree. There is an assertion below to check if the block is marked. But the original code does not mark the block if we forward the object. I don't know how it works. We would better check with @wenyuzhao.

Reply:
Blocks are marked when they're allocated from the page-resource or poped out from the reusable blocks queue.  https://github.com/mmtk/mmtk-core/blob/master/src/policy/immix/block.rs#L158-L162 . The store of the mark byte only happens once per block. The `set_state` here is unnecessary and also less efficient. We should remove this.

Reply:
Thanks @wenyuzhao  for pointing to the code. The assertion failure actually failed for sticky immix. I think the following is the problem.

We have two immix allocators in the worker, one used for defrag (`copy` is `true`), and the other one for normal copy (`copy` is `false`)
https://github.com/mmtk/mmtk-core/blob/106fc71472e0d960e78da977ef2d43d54bbc5113/src/policy/immix/immixspace.rs#L803-L806

For a defrag immix GC, we use the defrag allocator, and the defrag allocator will initiate the block as marked.
https://github.com/mmtk/mmtk-core/blob/106fc71472e0d960e78da977ef2d43d54bbc5113/src/policy/immix/immixspace.rs#L776-L780
https://github.com/mmtk/mmtk-core/blob/106fc71472e0d960e78da977ef2d43d54bbc5113/src/policy/immix/block.rs#L158-L162

For a fast immix GC, we use the normal allocator which does not initiate the blocks as marked, but we use `trace_object_without_moving()` which marks blocks when we trace objects.
https://github.com/mmtk/mmtk-core/blob/106fc71472e0d960e78da977ef2d43d54bbc5113/src/policy/immix/immixspace.rs#L455

So in short, for fast GC, we mark each block at tracing, and for defrag GC, we initiate each block as marked. It is an optimization so we do not need to mark objects if they are copied in the GC.

For sticky immix, this no longer works. A nursery GC in sticky immix is not a defrag GC (so blocks are not initiated as marked), but it may move objects (so it uses `trace_object_with_opportunistic_copying()` which does not mark blocks if an object is copied). So there are some blocks that are not marked if we do not add the line, and we will see the assertion failure.

Reply:
I will think about what to do for this.

Reply:
I checked the Java MMTk. And it seems we do not need two different immix copying allocators for most immix plans (only generational immix needs two copying allocators). So I renamed the current `ImmixCopyContext` (with two copying allocators) to `ImmixHybridCopyContext`, and this is used by generational immix. I also added a `ImmixCopyContext` (with one copying allocator), and this is used by immix and sticky immix. In this case, we no longer need the extra block marking in `trace_object`. I will measure performance again and post here.

Reply:
That's not quite right. StickyImmix uses `nurseryCopy` as well as `copy`. Note that it extends `ImmixCollector` so it already has the `ImmixAllocator copy` field.

Reply:
I didn't find where `nurseryCopy` is used. Is it used anywhere? It inherits `ImmixCollector`, but does not override `allocCopy`.

Reply:
Huh -- very odd... You're right. It's initialized but not used. `nurseryCopy.reset()` is called but nothing is ever alloced in it. I'll check git logs to see if it was ever used

Reply:
It was never used from the initial commit which added Immix (https://github.com/mmtk/jikesrvm/commit/f10ba059ebdb801cc66e2cb6552a6f4a3a0f193d). How odd.

Reply:
> I checked the Java MMTk. And it seems we do not need two different immix copying allocators for most immix plans (only generational immix needs two copying allocators). So I renamed the current `ImmixCopyContext` (with two copying allocators) to `ImmixHybridCopyContext`, and this is used by generational immix. I also added a `ImmixCopyContext` (with one copying allocator), and this is used by immix and sticky immix. In this case, we no longer need the extra block marking in `trace_object`. I will measure performance again and post here.

The results are [here](https://squirrel.anu.edu.au/plotty/yilin/mmtk/#0|ferret-2023-02-21-Tue-004454&benchmark^build^invocation^iteration&GC^time^time.other^time.stw&|10&iteration^1^2|20&1^invocation|50&GC^1^1^+inf|30&1&benchmark&build;jdk-mmtk-master-ix|41&Histogram%20(with%20CI)^build^benchmark&). I am still running it. For what I have now, the performance looks similar to my previous run (https://github.com/mmtk/mmtk-core/pull/741#issuecomment-1432284832).